File: python/ch20_agent/agent_ch20/core/profiles.py
"""固定章节能力快照。

Java 对照：`ChapterProfile` 类似不可变配置 record，P01-P20 类似预定义单例常量。
组合根使用对象身份判断（`is`），调用方不能临时拼一个同字段对象来冒充正式章节。

第 20 章把原来逐章手写的能力集合改成"增量表 + 累计推导"：
每章只声明自己新增的能力，`_PROFILES` 再把前缀累加成完整档案。
这样"第 N 章档案必须是第 N-1 章的严格超集"由数据结构保证，而不是靠人工核对。
"""

from dataclasses import dataclass
from typing import Literal

Capability = Literal[
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
    "skills",
    "artifacts",
    "compaction",
    "memory",
    "dynamic_prompt",
    "recovery",
    "task_dag_json",
    "background",
    "cron",
    "teammate",
    "mailbox",
    "protocol",
    "plan_gate",
    # task_dag_sqlite 与 work_stealing 表示带租约的 SQLite 任务图与去中心化认领。
    "task_dag_sqlite",
    "work_stealing",
    # worktree 表示受控 Git 工作树隔离，mcp 表示可动态发布和撤销的远程 MCP 工具边界。
    "worktree",
    "mcp",
    # full_harness 标记 P20 完整能力集，不新增独立运行时。
    "full_harness",
]


@dataclass(frozen=True, slots=True)
class ChapterProfile:
    """章节号及其允许装配的能力白名单。

    字段说明：
    - `chapter`：章节号，组合根用它填充 Prompt 的静态 context；
    - `capabilities`：该章允许启用的能力集合，`frozenset` 保证构造后不可增删。

    Java 对照：等价于一个只有 getter 的不可变 record，`frozenset` 类似
    `Set.copyOf(...)` 返回的只读集合。
    """

    chapter: int
    capabilities: frozenset[Capability]


# 每章只声明"本章新增"的能力；完整档案由前缀累加得到。
# Java 对照：类似 `List<List<String>>` 常量表，元素顺序即章节顺序。
_PROFILE_DELTAS: tuple[tuple[Capability, ...], ...] = (
    ("loop", "powershell"),
    ("tool_registry", "files"),
    ("policy",),
    ("hooks",),
    ("todo",),
    ("subagent",),
    ("skills",),
    ("artifacts", "compaction"),
    ("memory",),
    ("dynamic_prompt",),
    ("recovery",),
    ("task_dag_json",),
    ("background",),
    ("cron",),
    ("teammate", "mailbox"),
    ("protocol", "plan_gate"),
    ("task_dag_sqlite", "work_stealing"),
    ("worktree",),
    ("mcp",),
    # full_harness 是 P20 的标记能力：不新增具体运行时，只声明前十九章能力全部生效。
    ("full_harness",),
)


def _build_profiles() -> tuple[ChapterProfile, ...]:
    """把增量表累加成每章的完整档案。

    这里用普通 for 循环而不是推导式，因为要维护一个跨轮次累积的 `accumulated` 列表；
    Java 对照：等价于在循环里对 `Set` 反复 `addAll` 并每轮快照一次。
    """
    profiles: list[ChapterProfile] = []
    accumulated: list[Capability] = []
    for index, delta in enumerate(_PROFILE_DELTAS):
        # extend 就地追加本章增量，因此下一轮天然包含所有历史能力。
        accumulated.extend(delta)
        profiles.append(ChapterProfile(index + 1, frozenset(accumulated)))
    # tuple(...) 冻结外层序列，避免调用方替换某一章的档案对象。
    return tuple(profiles)


_PROFILES = _build_profiles()


def _profile_at(chapter: int) -> ChapterProfile:
    """按章节号取固定档案；越界视为编程错误而不是返回空能力集。"""
    if not 1 <= chapter <= len(_PROFILES):
        raise ValueError(f"第 {chapter} 章尚未迁移为 Python")
    return _PROFILES[chapter - 1]


P01 = _profile_at(1)
P02 = _profile_at(2)
P03 = _profile_at(3)
P04 = _profile_at(4)
P05 = _profile_at(5)
P06 = _profile_at(6)
P07 = _profile_at(7)
P08 = _profile_at(8)
P09 = _profile_at(9)
P10 = _profile_at(10)
P11 = _profile_at(11)
P12 = _profile_at(12)
P13 = _profile_at(13)
P14 = _profile_at(14)
P15 = _profile_at(15)
P16 = _profile_at(16)
P17 = _profile_at(17)
P18 = _profile_at(18)
P19 = _profile_at(19)
# P20 是完整 Harness 档案，统一启用前十九章已经验证的累计能力。
P20 = _profile_at(20)


def profile_for_chapter(chapter: int) -> ChapterProfile:
    """只返回模块内固定单例；越界或非整数输入立即失败。

    `isinstance(chapter, bool)` 需要单独排除，因为 Python 的 `bool` 是 `int` 子类，
    `True` 会被 `isinstance(True, int)` 判为真。Java 中 `boolean` 与 `int` 无继承关系，
    不需要这一步。
    """
    if isinstance(chapter, bool) or not isinstance(chapter, int) or not 1 <= chapter <= 20:
        raise ValueError("chapter 必须是 1 到 20 的整数")
    return _profile_at(chapter)

File: typescript/chapters/ch18/src/core/profiles.ts
// 章节能力清单：每个 profile 都是冻结单例，组合根据此决定启用的运行时边界。
export type Capability =
  | "loop"
  | "powershell"
  | "tool_registry"
  | "files"
  | "policy"
  | "hooks"
  | "todo"
  | "subagent"
  | "skills"
  | "artifacts"
  | "compaction"
  | "memory"
  | "dynamic_prompt"
  | "recovery"
  | "task_dag_json"
  | "background"
  | "cron"
  | "teammate"
  | "mailbox"
  // protocol 与 plan_gate 共同表示结构化审批消息和副作用执行门控。
  | "protocol"
  | "plan_gate"
  // task_dag_sqlite 与 work_stealing 表示带租约的 SQLite 任务图与去中心化认领。
  | "task_dag_sqlite"
  | "work_stealing"
  // worktree 启用受管 Git 工作树与动态 ToolContextProvider。
  | "worktree";

export interface ChapterProfile {
  // chapter 与能力集共同构成 profileForChapter 返回的固定单例。
  readonly chapter: number;
  readonly capabilities: ReadonlySet<Capability>;
}

class CapabilitySet implements ReadonlySet<Capability> {
  // 包装原生 Set，只暴露 ReadonlySet 契约。
  readonly #values: Set<Capability>;

  constructor(values: readonly Capability[]) {
    // 复制输入，外部修改不会改变已构建 profile。
    this.#values = new Set(values);
  }

  get size(): number {
    // 返回去重后的能力数量。
    return this.#values.size;
  }

  has(value: Capability): boolean {
    // 组合根通过能力位选择依赖，不散落章节号判断。
    return this.#values.has(value);
  }

  entries(): SetIterator<[Capability, Capability]> {
    // 迭代行为与原生 Set 一致。
    return this.#values.entries();
  }

  keys(): SetIterator<Capability> {
    // 不暴露内部可变 Set。
    return this.#values.keys();
  }

  values(): SetIterator<Capability> {
    // 保留能力构建顺序，便于下一章累计展开。
    return this.#values.values();
  }

  forEach(
    callbackfn: (value: Capability, value2: Capability, set: ReadonlySet<Capability>) => void,
    thisArg?: unknown,
  ): void {
    // 第三个参数传只读包装本身。
    this.#values.forEach((value) => {
      callbackfn.call(thisArg, value, value, this);
    });
  }

  [Symbol.iterator](): SetIterator<Capability> {
    // 支持扩展运算符构建累计能力集。
    return this.values();
  }
}

export const P01: ChapterProfile = Object.freeze({
  chapter: 1,
  capabilities: new CapabilitySet(["loop", "powershell"]),
});

export const P02: ChapterProfile = Object.freeze({
  chapter: 2,
  capabilities: new CapabilitySet(["loop", "powershell", "tool_registry", "files"]),
});

export const P03: ChapterProfile = Object.freeze({
  chapter: 3,
  capabilities: new CapabilitySet(["loop", "powershell", "tool_registry", "files", "policy"]),
});

export const P04: ChapterProfile = Object.freeze({
  chapter: 4,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
  ]),
});

export const P05: ChapterProfile = Object.freeze({
  chapter: 5,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
  ]),
});

export const P06: ChapterProfile = Object.freeze({
  chapter: 6,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
  ]),
});

export const P07: ChapterProfile = Object.freeze({
  chapter: 7,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
    "skills",
  ]),
});

export const P08: ChapterProfile = Object.freeze({
  chapter: 8,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
    "skills",
    "artifacts",
    "compaction",
  ]),
});

export const P09: ChapterProfile = Object.freeze({
  chapter: 9,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
    "skills",
    "artifacts",
    "compaction",
    "memory",
  ]),
});

export const P10: ChapterProfile = Object.freeze({
  chapter: 10,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
    "skills",
    "artifacts",
    "compaction",
    "memory",
    "dynamic_prompt",
  ]),
});

export const P11: ChapterProfile = Object.freeze({
  chapter: 11,
  capabilities: new CapabilitySet([
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
    "skills",
    "artifacts",
    "compaction",
    "memory",
    "dynamic_prompt",
    "recovery",
  ]),
});

export const P12: ChapterProfile = Object.freeze({
  chapter: 12,
  capabilities: new CapabilitySet([...P11.capabilities, "task_dag_json"]),
});

export const P13: ChapterProfile = Object.freeze({
  chapter: 13,
  capabilities: new CapabilitySet([...P12.capabilities, "background"]),
});

export const P14: ChapterProfile = Object.freeze({
  chapter: 14,
  capabilities: new CapabilitySet([...P13.capabilities, "cron"]),
});

export const P15: ChapterProfile = Object.freeze({
  chapter: 15,
  capabilities: new CapabilitySet([...P14.capabilities, "teammate", "mailbox"]),
});

export const P16: ChapterProfile = Object.freeze({
  chapter: 16,
  capabilities: new CapabilitySet([...P15.capabilities, "protocol", "plan_gate"]),
});

// P17 在 P16 基础上追加 SQLite 任务图与 work stealing，让 Lead、子代理和队友共享同一认领路径。
export const P17: ChapterProfile = Object.freeze({
  chapter: 17,
  capabilities: new CapabilitySet([...P16.capabilities, "task_dag_sqlite", "work_stealing"]),
});

// P18 在 SQLite 工作窃取之上加入受控 Git worktree 隔离。
export const P18: ChapterProfile = Object.freeze({
  chapter: 18,
  capabilities: new CapabilitySet([...P17.capabilities, "worktree"]),
});

export function profileForChapter(chapter: number): ChapterProfile {
  if (chapter === 1) {
    return P01;
  }
  if (chapter === 2) {
    return P02;
  }
  if (chapter === 3) {
    return P03;
  }
  if (chapter === 4) {
    return P04;
  }
  if (chapter === 5) {
    return P05;
  }
  if (chapter === 6) {
    return P06;
  }
  if (chapter === 7) {
    return P07;
  }
  if (chapter === 8) {
    return P08;
  }
  if (chapter === 9) {
    return P09;
  }
  if (chapter === 10) {
    return P10;
  }
  if (chapter === 11) {
    return P11;
  }
  if (chapter === 12) {
    return P12;
  }
  if (chapter === 13) {
    return P13;
  }
  if (chapter === 14) {
    return P14;
  }
  if (chapter === 15) {
    return P15;
  }
  if (chapter === 16) {
    return P16;
  }
  if (chapter === 17) {
    return P17;
  }
  if (chapter === 18) {
    return P18;
  }
  if (!Number.isInteger(chapter) || chapter < 1 || chapter > 20) {
    throw new Error("chapter must be an integer from 1 to 20");
  }
  throw new Error(`Chapter ${chapter} has not been migrated to TypeScript yet`);
}
